### PR TITLE
Range label customization

### PR DESCRIPTION
**Pull request by snorham**
Changes the way in which coordinates for range (previously extreme) labels are calculated and allows for customizable placement of said labels.

### CODE REVIEW

**Comment by snorham** [comment]
I may have handled the label drawing with some wacky math and/or a faulty assumption about what `minimumBackgroundArcAngle` and `maximumBackgroundArcAngle` are meant to represent.  Everything seems to go haywire when I set new values to the aforementioned variables, but this "unexpected" behavior seems consistent with what I see on our `master` branch.  The label placement works fine otherwise.

**Comment by kyleofori** [comment]
It looks like there's still a merge conflict; don't know what happened there.

File: doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java [hunk 1]
```diff
@@ -110,9 +133,9 @@ private static void initDefaultForegroundArcPaint() {
 
     /**
      * Aspect ratio of child view, such that
-     *
-     *     w = LAMBDA h
-     *
+     * <p/>
```
Comment (kyleofori) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java:138]: I didn't know you could do this! Is there any way to see what it looks like on a website or in official documentation?

**Comment by kyleofori** [comment]
It looks like there's still a merge conflict; don't know what happened there.

File: doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java (hunk 1 above)
Comment (snorham) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java:138]: I actually didn't add that.  It must have happened when I automatically formatted this class.

**Comment by snorham** [comment]
conflict clobbered

File: doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java (hunk 1 above)
Comment (kyleofori) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java:138]: Fancy.

File: doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java [hunk 2]
```diff
@@ -140,7 +163,8 @@ private void init() {
         DEFAULT_FOREGROUND_ARC_PAINT.setStrokeWidth(dpToPx(DEFAULT_ARC_STROKE_WIDTH_DP));
 
         DEFAULT_CURRENT_VALUE_LABEL_PAINT.setTextSize(spToPx(DEFAULT_CURRENT_VALUE_LABEL_TEXT_SIZE_SP));
-        DEFAULT_EXTREME_VALUE_LABEL_PAINT.setTextSize(spToPx(DEFAULT_EXTREME_VALUE_LABEL_TEXT_SIZE_SP));
+        DEFAULT_MINIMUM_LABEL_PAINT.setTextSize(spToPx(DEFAULT_RANGE_LABEL_TEXT_SIZE_SP));
```
Comment (kyleofori) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java]: Stu mentioned that it's useful to put units for measurements like this, so... `setTextSizePx()`?

Comment (snorham) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java]: Are you referring to `DEFAULT_MINIMUM_LABEL_PAINT.setTextSize()`?  If so, that is a method of the `Paint` class.

Comment (kyleofori) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java]: Ah. Nothing to be done there, then.

File: doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java [hunk 3]
```diff
@@ -371,31 +397,89 @@ public void setCurrentValueLabelTextColor(@ColorInt final int textColor) {
...
+        newMaxPaint.setTypeface(typeface);
+        customMaximumLabelPaint = newMaxPaint;
         invalidate();
     }
 
-    public void setExtremeValueLabelTextSizePx(final float textSizePx) {
-        final Paint newPaint = new Paint(getExtremeLabelTextPaint());
-        newPaint.setTextSize(textSizePx);
-        customExtremeValueLabelPaint = newPaint;
+    public void alignRangeLabelTextInward() {
```
Comment (kyleofori) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java]: What are your thoughts on an `alignRangeLabelText()` or `setRangeLabelTextAlignment()` method that takes an enum as a parameter? We could use `INWARD`, `OUTWARD` and `CENTERED` as options.

Comment (kyleofori) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java]: The names could use some work, but this is what I had in mind:

```
public void alignRangeLabelText(final Alignment alignment) {
    switch (alignment) {
        case INWARD:
            setRangeLabelTextAlignment(Paint.Align.LEFT, Paint.Align.RIGHT);
            break;
        case OUTWARD:
            setRangeLabelTextAlignment(Paint.Align.RIGHT, Paint.Align.LEFT);
            break;
        case CENTERED:
        default:
            setRangeLabelTextAlignment(Paint.Align.CENTER, Paint.Align.CENTER);
            break;
    }
}

private void setRangeLabelTextAlignment(final Paint.Align minAlign, final Paint.Align maxAlign) {
    final Paint newMinPaint = new Paint(getMinimumLabelTextPaint());
    newMinPaint.setTextAlign(minAlign);
    customMinimumLabelPaint = newMinPaint;

    final Paint newMaxPaint = new Paint(getMaximumLabelTextPaint());
    newMaxPaint.setTextAlign(maxAlign);
    customMaximumLabelPaint = newMaxPaint;
    invalidate();
}
```

Comment (snorham) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java]: I'm game.  I'll add it.

**Comment by snorham** [comment]
Ok.  I scrapped a lot of a code centered around how the arc is being drawn and how a developer would create custom arcs (as well as custom label placement) with the doomrainbow library.  Arcs are now made with a starting point and a sweep angle in mind, just like with `Canvas.drawArc()`.  Rainbows no longer have "minimum" or "maximum" endpoints, as those words could be confusing if you were to "start" the arc at a positive number and then "end" it with a negative number.  Arcs can now sweep in both directions (regardless of starting and ending values) and start at any point on a circle.  Within reason, their "start" and "end" labels should look OK no matter how the arc is shaped.

File: doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/AngleUtils.java [hunk 4]
```diff
@@ -7,12 +7,35 @@ public static float getRadiusCosineCoefficient(final float valuePositionInDegree
...
+            final float exteriorEndValue,
+            final float exteriorStartAngle,
+            final float exteriorSweepAngle) {
+
+        if (exteriorStartValue < exteriorEndValue) {
+            if(value <= exteriorStartValue) {
+                return exteriorStartAngle;
+            } else if (value >= exteriorEndValue) {
+                return exteriorStartAngle + exteriorSweepAngle;
+            } else {
```
Comment (kyleofori) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/AngleUtils.java]: As Stu once told me, [don't repeat yourself](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). 
The code in this else block could be made into its own method so it doesn't have to be repeated.

**Comment by kyleofori** [comment]
Whooooaaa this is nice! Cool that it works smoothly without having to add a bunch of new methods.

**Comment by snorham** [comment]
@kyleofori check out the latest commit.  I refactored the method you commented on.  I also moved it into RainbowView.  The large number of arguments seemed to indicate that is where it belongs.  It's either that or pass the entire RainbowView.

File: doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java [hunk 5]
```diff
@@ -630,14 +754,10 @@ public void onAnimationUpdate(final ValueAnimator valueAnimator) {
...
-        return maximumValue - minimumValue;
-    }
-
-    private float getBackgroundArcAngleRangeLength() {
-        return maximumBackgroundArcAngle - minimumBackgroundArcAngle;
+        return endValue - startValue;
     }
 
-    private float spToPx(final float sp) {
+     private float spToPx(final float sp) {
```
Comment (stkent) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RainbowView.java]: extra space should be removed

File: doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RangeLabelAlignment.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package com.detroitlabs.kyleofori.doomrainbow;
+
+public enum RangeLabelAlignment {
+    INWARD, OUTWARD, CENTERED
```
Comment (stkent) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RangeLabelAlignment.java:4]: these might be confusing if the user selected certain start/sweep angle combos, but i don't see a good way to work around it at the moment...

Comment (snorham) [inline doomrainbow/src/main/java/com/detroitlabs/kyleofori/doomrainbow/RangeLabelAlignment.java:4]: Me neither... unless we just allow the labels to be aligned left,right, and center individually.

**Pull request merged**